### PR TITLE
[API Compatibility] Support param `optimizer` for `lr_scheduler`

### PR DESCRIPTION
**Pull request by algorithm1832**
### PR Category
User Experience

### PR Types
Improvements

### Description

- Add alias `optimizer` for `learning_rate`
- Add learning rate setting and scheduler assignment when `optimizer` is passed
- Add param alias for `ReduceOnPlateau`
- Add corresponding tests

**Used AI Studio**

Related APIs:
```
LRScheduler
ExponentialDecay
CosineAnnealingDecay
CosineAnnealingWarmRestarts
MultiStepDecay
ReduceOnPlateau
StepDecay
```

### 是否引起精度变化
否

### CODE REVIEW

File: python/paddle/optimizer/lr.py [hunk 1]
```diff
@@ -144,20 +145,26 @@ class LRScheduler:
...
-        if not isinstance(learning_rate, (float, int)):
+        if isinstance(learning_rate, (float, int)):
+            if learning_rate < 0:
+                raise ValueError(f"Invalid learning rate: {learning_rate}")
+            self.base_lr = float(learning_rate)
+            self.last_lr = float(learning_rate)
+        elif isinstance(learning_rate, paddle.optimizer.Optimizer):
+            self.base_lr = learning_rate.get_lr()
+            self.last_lr = learning_rate.get_lr()
+            learning_rate.set_lr_scheduler(self)
```
Comment (risemeup1111) [inline python/paddle/optimizer/lr.py]: P2 **优先级：P2**

这里直接调用 `optimizer.set_lr_scheduler(self)`，但该 API 被 `@framework.dygraph_only` 限制；因此在 static/PIR 模式下，新引入的 `optimizer` 参数会在构造 scheduler 时抛 `AssertionError`。这些 lr scheduler 现有文档和测试都覆盖 static 图用法，PR 也没有声明 `optimizer` 只支持动态图，会形成不一致的兼容行为。

建议把 optimizer 分支拆成：动态图继续调用 `set_lr_scheduler`；静态图在 optimizer 尚未创建 lr var 时把 `_learning_rate` 设为当前 scheduler（或显式拒绝并在文档/测试中说明 dynamic-only）。同时补一个 static 覆盖，例如 `Adam(learning_rate=0.1)` 后 `StepDecay(optimizer=adam)` 再 `minimize`。实现形状可以类似：

```python
elif isinstance(learning_rate, paddle.optimizer.Optimizer):
    self.base_lr = learning_rate.get_lr()
    self.last_lr = learning_rate.get_lr()
    if in_dygraph_mode():
        learning_rate.set_lr_scheduler(self)
    elif not learning_rate._learning_rate_map:
        learning_rate._learning_rate = self
    else:
        raise RuntimeError(
            "optimizer can only be passed before the learning rate variable is created in static graph mode."
        )
```

File: python/paddle/optimizer/lr.py [hunk 2]
```diff
@@ -1595,8 +1602,17 @@ def __init__(
...
         # Can not call Parent __init__, so implement here.
-        self.base_lr = float(learning_rate)
-        self.last_lr = float(learning_rate)
+        if isinstance(learning_rate, (float, int)):
+            self.base_lr = float(learning_rate)
+            self.last_lr = float(learning_rate)
+        elif isinstance(learning_rate, paddle.optimizer.Optimizer):
+            self.base_lr = learning_rate.get_lr()
+            self.last_lr = learning_rate.get_lr()
+            learning_rate.set_lr_scheduler(self)
```
Comment (risemeup1111) [inline python/paddle/optimizer/lr.py]: P2 **优先级：P2**

`ReduceOnPlateau` 这里复制了基类的 optimizer 分支，也直接调用了 `set_lr_scheduler()`，所以同样只在动态图可用；static/PIR 模式传 `optimizer=` 会在构造时失败。由于 `ReduceOnPlateau` 没有调用父类 `__init__`，即使修复基类路径，这里也需要同步处理并补 static 覆盖。

建议和基类保持一致，例如：

```python
elif isinstance(learning_rate, paddle.optimizer.Optimizer):
    self.base_lr = learning_rate.get_lr()
    self.last_lr = learning_rate.get_lr()
    if in_dygraph_mode():
        learning_rate.set_lr_scheduler(self)
    elif not learning_rate._learning_rate_map:
        learning_rate._learning_rate = self
    else:
        raise RuntimeError(
            "optimizer can only be passed before the learning rate variable is created in static graph mode."
        )
```

**Review by risemeup1111 — COMMENTED** [review]
未发现需要阻塞合入的问题。已在行级评论中标出 `optimizer` 参数在 static/PIR 路径下的兼容性缺口，建议补齐实现和对应测试后再合入。当前 CI 仍有任务运行中，请以最终 CI 结果为准。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: python/paddle/optimizer/lr.py [hunk 3]
```diff
@@ -144,20 +145,26 @@ class LRScheduler:
     last_epoch: int
     verbose: bool
 
+    @param_one_alias(["learning_rate", "optimizer"])
     def __init__(
```
Comment (zhwesky2010) [inline python/paddle/optimizer/lr.py:168]: 这里写一个装饰器来处理这个地方的问题吧，直接将optimizer作为learning_rate的别名会显得很奇怪

上面再加两个overload注解

下同

**Comment by zhwesky2010** [comment]
@algorithm1832 这个PR修改一下

File: python/paddle/optimizer/lr.py [hunk 4]
```diff
@@ -144,20 +148,42 @@ class LRScheduler:
...
+        self,
+        optimizer: paddle.optimizer.Optimizer,
+        last_epoch: int = -1,
+        verbose: bool = False,
+    ) -> None: ...
+
+    @lr_scheduler_optimizer_decorator()
+    def __init__(
+        self,
+        learning_rate: float | paddle.optimizer.Optimizer = 0.1,
```
Comment (zhwesky2010) [inline python/paddle/optimizer/lr.py]: 这里的签名不用改，显得比较奇怪

File: python/paddle/optimizer/lr.py [hunk 5]
```diff
@@ -144,20 +148,42 @@ class LRScheduler:
...
+        last_epoch: int = -1,
+        verbose: bool = False,
     ) -> None:
-        if not isinstance(learning_rate, (float, int)):
+        if isinstance(learning_rate, (float, int)):
```
Comment (zhwesky2010) [inline python/paddle/optimizer/lr.py]: 这些逻辑不应该都进装饰器了吗，就不用在这处理了。这个init保持和原来不变

File: python/paddle/optimizer/lr.py [hunk 6]
```diff
@@ -2581,7 +2727,29 @@ def __init__(
...
+        self,
+        optimizer: paddle.optimizer.Optimizer,
+        T_0: int,
+        T_mult: int = 1,
+        eta_min: float = 0,
+        last_epoch: int = -1,
+        verbose: bool = False,
+    ) -> None: ...
+
+    @lr_scheduler_optimizer_decorator()
```
Comment (zhwesky2010) [inline python/paddle/optimizer/lr.py]: 命名：lr_scheduler_decorator

**Review by zhwesky2010 — COMMENTED** [review]
这些逻辑都放到decrator里来，init不要动，init注解overload就行

File: test/legacy_test/test_lr_scheduler.py [hunk 7]
```diff
@@ -1332,6 +1332,84 @@ def test_pir_linear_warmup_lr(self):
...
+            learning_rate=0.5, parameters=linear.parameters()
+        )
+        scheduler = paddle.optimizer.lr.MultiStepDecay(
+            optimizer=sgd, milestones=[2, 4, 6]
+        )
+        self.assertEqual(scheduler.base_lr, sgd.get_lr())
+        self.assertIs(sgd._learning_rate, scheduler)
+        paddle.enable_static()
+
+    def test_reduce_on_plateau(self):
```
Comment (zhwesky2010) [inline test/legacy_test/test_lr_scheduler.py:1423]: 要像这样测下简易的网络

Image

File: python/paddle/utils/decorator_utils.py [hunk 8]
```diff
@@ -1189,36 +1189,37 @@ def wrapper(*args: _InputT.args, **kwargs: _InputT.kwargs) -> _RetT:
...
-            # args[2] is wrap, use torch signature
-            if len(args) >= 3 and isinstance(args[2], bool):
-                kwargs["wrap"] = args[2]
-                if len(args) > 3:
-                    raise TypeError(
-                        "fill_diagonal_() received too many arguments"
-                    )
-                args = (args[0], args[1])
+            if opt is not None:
+                opt.set_lr_scheduler(args[0])
```
Comment (zhwesky2010) [inline python/paddle/utils/decorator_utils.py:1222]: 这一块是在 init 之前就set了，这个会有问题不

**Review by zhwesky2010 — DISMISSED** [review]

**Comment by algorithm1832** [comment]
CI-H失败原因为超时被kill

**Review by zhwesky2010 — APPROVED** [review]
LGTM
